main.go

```diff
@@ -609,6 +611,81 @@ func newBufPool(sz int) httputil.BufferPool {
 	}}
 }
 
+type byLastOctetValue []net.IP
+
+func (n byLastOctetValue) Len() int      { return len(n) }
+func (n byLastOctetValue) Swap(i, j int) { n[i], n[j] = n[j], n[i] }
+func (n byLastOctetValue) Less(i, j int) bool {
+	// This case is needed when all ips in the list
+	// have same last octets, Following just ensures that
+	// 127.0.0.1 is moved to the end of the list.
+	if n[i].IsLoopback() {
+		return false
+	}
+	if n[j].IsLoopback() {
+		return true
+	}
+	return []byte(n[i].To4())[3] > []byte(n[j].To4())[3]
+}
+
+// sortIPs - sort ips based on higher octects.
+// The logic to sort by last octet is implemented to
+// prefer CIDRs with higher octects, this in-turn skips the
+// localhost/loopback address to be not preferred as the
+// first ip on the list. Subsequently this list helps us print
+// a user friendly message with appropriate values.
+func sortIPs(ipList []string) []string {
+	if len(ipList) == 1 {
+		return ipList
+	}
+
+	var ipV4s []net.IP
+	var nonIPs []string
+	for _, ip := range ipList {
+		nip := net.ParseIP(ip)
+		if nip != nil {
+			ipV4s = append(ipV4s, nip)
+		} else {
+			nonIPs = append(nonIPs, ip)
+		}
+	}
+
+	sort.Sort(byLastOctetValue(ipV4s))
+
+	var ips []string
+	for _, ip := range ipV4s {
+		ips = append(ips, ip.String())
+	}
+
+	return append(nonIPs, ips...)
+}
+
+func getPublicIP() string {
+	if globalClientIP != "" {
+		return globalClientIP
+	}
+	var IPs []string
+	addrs, _ := net.InterfaceAddrs()
+	for _, addr := range addrs {
+		ipNet, ok := addr.(*net.IPNet)
+		if ok && !ipNet.IP.IsLoopback() && ipNet.IP.To4() != nil {
+			IPs = append(IPs, ipNet.IP.String())
+		}
+	}
+	globalClientIP = sortIPs(IPs)[0] // There would be at least one entry
```

Comment:
this kind of looks wrong, you are editing global value for an active caller during proxy which is racy.

Reply:
So this should locked or done at different stage?

Reply:
No no globalClientIP is the same for x-forwarded-for and x-real-ip - x-forwarded-host is always pointing to r.Host we wouldn't be doing `getHost()` for that. 

WE need to make sure to set the right r.Host otherwise the signature will fail. We have to keep the r.Host same as what the application used in this case it will be 127.0.0.1 or whatever. 

You can test it out with

```
./sidekick --health-path=/minio/health/ready https://play.min.io
```